extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTools.java

```diff
@@ -456,15 +464,124 @@ private static Map<String, ParquetTableWriter.GroupingColumnWritingInfo> groupin
+        }
+        final String[] partitioningColNames = partitioningColumns.stream()
+                .map(ColumnDefinition::getName)
+                .toArray(String[]::new);
+        final PartitionedTable partitionedTable = sourceTable.partitionBy(partitioningColNames);
+        writeKeyValuePartitionedTable(partitionedTable, destinationDir, baseName, writeInstructions);
+    }
+
+    /**
+     * Write a partitioned tables to disk in parquet format with all the {@link PartitionedTable#keyColumnNames() key
+     * columns} as "key=value" format in a nested directory structure. To generate the partitioned table, users can call
+     * {@link Table#partitionBy(String...) partitionBy} on the required columns.
+     *
+     * @param partitionedTable The partitioned table to write
+     * @param destinationDir The destination directory to store partitioned data in. Non-existing directories are
+     *        created.
+     * @param baseName The base name for the individual partitioned tables. For example, a base name of "table" will
+     *        result in files named "partition1/table.parquet", "partition2/table.parquet", etc.
+     * @param writeInstructions Write instructions for customizations while writing
+     */
+    public static void writeKeyValuePartitionedTable(@NotNull final PartitionedTable partitionedTable,
+            @NotNull final File destinationDir,
+            @NotNull final String baseName,
+            @NotNull final ParquetInstructions writeInstructions) {
+        final String[] partitioningColumnNames = partitionedTable.keyColumnNames().toArray(String[]::new);
+        final Table keyTable;
+        if (partitionedTable.uniqueKeys()) {
+            keyTable = partitionedTable.table().view(partitioningColumnNames);
+        } else {
+            keyTable = partitionedTable.table().selectDistinct(partitioningColumnNames);
+        }
+        final Collection<Table> constituentTables = new ArrayList<>();
+        final Collection<File> destinations = new ArrayList<>();
+        keyTable.getRowSet().forAllRowKeys(key -> {
+            final Object[] keyValues = Arrays.stream(partitioningColumnNames)
+                    .map(keyTable::getColumnSource)
+                    .map(colSource -> colSource.get(key))
+                    .toArray();
+            final StringBuilder partitionTableRelativePath = new StringBuilder();
+            for (int i = 0; i < partitioningColumnNames.length; i++) {
+                partitionTableRelativePath.append(File.separator)
+                        .append(partitioningColumnNames[i]).append("=").append(keyValues[i]);
+            }
+            partitionTableRelativePath.append(File.separator).append(baseName).append(".parquet");
+            destinations.add(new File(destinationDir, partitionTableRelativePath.toString()));
+            constituentTables.add(partitionedTable.constituentFor(keyValues));
```

Comment:
This may break if we hit the `!uniqueKeys()` case.
Maybe we want to just "assume the worst" and use:
```
        Table withGroupConstituents = partitionedTable.table().groupBy(partitionedTable.keyColumnNames().toArray(String[]::new));
```
So we can operate on `ObjectVector<? extends Table>`.

We should insist on a UUID or an index in the name template if `!uniqueKeys()`.